libvast/src/directory.cpp

```diff
@@ -108,23 +109,18 @@ const path& directory::path() const {
 }
 
 size_t recursive_size(const vast::directory& dir) {
-  size_t size = 0;
-  std::vector<vast::directory> stack{dir};
-  while (!stack.empty()) {
-    auto dir = std::move(stack.back());
-    stack.pop_back();
-    for (auto file : dir) {
-      if (file.is_regular_file()) {
-        if (auto sz = vast::file_size(file)) {
-          VAST_TRACE_SCOPE("{} += {}", file, *sz);
-          size += *sz;
-        }
-      } else if (file.is_directory()) {
-        stack.push_back(vast::directory{file});
-      }
+  size_t total_size = 0;
+
+  const auto p = std::filesystem::path{dir.path().str()};
+  for (const auto& f : std::filesystem::recursive_directory_iterator(p)) {
```

Comment:
This function can throw, so we should use the overload with a `std::error_code`. This means the function needs to return `expected<size_t>`:
```cpp
std::error_code ec;
auto dir = std::filesystem::recursive_directory_iterator(p, ec);
if (ec)
   return caf::make_error(ec); // TODO: discuss out how to render the error
for (const auto& f : dir)
   // ...
```

@tenzir/backend  can you flesh out this error-handling strawman?

Reply:
Valid point. I just switched to use the error code API for `std::filesystem` here. In the case there is an error, I use the message as returned from `std::error_code` to pass through in the `caf::error` type. This gives the caller the most information and can choose what to do with it.

I'm not sure in this case what we're looking to do in the error cases from the caller side in `disk_monitor.cpp`. See what I've done and let me know.

Reply:
Generally, there are three ways to handle errors in actor behaviors:

1. Log using `VAST_ERROR` and continue operation.
2. Return the error to the requesting actor (you can change the result type of the behavior lambda to `caf::result<void>` to be able to return errors.
3. Shut down the actor using `self->quit(error)`.

Now the disk monitor is running in a loop, so (2) doesn't quite make sense because it's requesting itself periodically to perform the check. (3) seems a bit harsh, as there's no need to shut down the disk monitor. So (1) it should be.

Reply:
If you want me to, I can push the necessary adaptions directly onto your branch.

Reply:
Thanks for the explanation for error handling.  Do you mind pushing the necessary changes for error handling in the `disk_monitor.cpp` loop?

Reply:
Can do, but might take me a few days—off of work until Monday.

Reply:
I cannot push to your branch, but here are the remaining changes I'd make to this PR:

```diff
diff --git a/libvast/src/system/disk_monitor.cpp b/libvast/src/system/disk_monitor.cpp
index 9b656a4d3..ac38bf2f7 100644
--- a/libvast/src/system/disk_monitor.cpp
+++ b/libvast/src/system/disk_monitor.cpp
@@ -62,7 +62,10 @@ disk_monitor(disk_monitor_actor::stateful_pointer<disk_monitor_state> self,
       // see noticeable overhead even on large-ish databases.
       // Nonetheless, if this becomes relevant we should switch to using
       // `inotify()` or similar to do real-time tracking of the db size.
-      if (const auto size = recursive_size(self->state.dbdir); size) {
+      if (const auto size = recursive_size(self->state.dbdir); !size) {
+        VAST_WARN("{} failed to calculate recursive size of {}: {}", self,
+                  self->state.dbdir, size.error());
+      } else {
         VAST_VERBOSE("{} checks db-directory of size {} bytes", self, *size);
         if (*size > self->state.high_water_mark && !self->state.purging) {
           self->state.purging = true;
@@ -115,19 +118,22 @@ disk_monitor(disk_monitor_actor::stateful_pointer<disk_monitor_state> self,
                         erased_ids)
               .then(
                 [=, sg = shared_guard](atom::done) {
-                  if (const auto sz = recursive_size(self->state.dbdir); sz) {
+                  if (const auto size = recursive_size(self->state.dbdir);
+                      !size) {
+                    VAST_WARN("{} failed to calculate recursive size of {}: {}",
+                              self, self->state.dbdir, size.error());
+                  } else {
                     VAST_VERBOSE("{} erased ids from index; {} bytes "
                                  "left on disk",
-                                 self, *sz);
-                    if (*sz > self->state.low_water_mark) {
+                                 self, *size);
+                    if (*size > self->state.low_water_mark) {
                       // Repeat until we're below the low water mark
                       self->send(self, atom::erase_v);
                     }
                   }
                 },
-                [=, sg = shared_guard](caf::error e) {
-                  VAST_WARN("{} failed to erase from archive: {}", self,
-                            render(e));
+                [=, sg = shared_guard](caf::error err) {
+                  VAST_WARN("{} failed to erase from archive: {}", self, err);
                 });
           },
           [=, sg = shared_guard](caf::error e) {
diff --git a/libvast/vast/directory.hpp b/libvast/vast/directory.hpp
index 8f6fae328..5552f37a8 100644
--- a/libvast/vast/directory.hpp
+++ b/libvast/vast/directory.hpp
@@ -13,18 +13,18 @@
 
 #pragma once
 
+#include "vast/fwd.hpp"
+
 #include "vast/config.hpp"
 #include "vast/defaults.hpp"
 #include "vast/detail/iterator.hpp"
 #include "vast/path.hpp"
 
-#include <caf/expected.hpp>
-
 #if VAST_POSIX
 #  include <dirent.h>
 #else
 
-#include <functional>
+#  include <functional>
 
 namespace vast {
 struct DIR;
```

Reply:
Just applied your diff -- thanks for sharing.  Not sure why you couldn't push to my branch on my fork. 🤔 

In any case, I applied the diff but kept the `<caf/expected.hpp>` include in there as we return `caf::expected` from `recursive_size` now. Did you intentionally mean to drop that include?

Reply:
Yes, the forward declaration from `fwd.hpp` suffices.